samples/read_aux.c

```diff
@@ -0,0 +1,210 @@
+/*  read_aux.c --  showcases the htslib api usage
+
+    Copyright (C) 2023 Genome Research Ltd.
+
+    Author: Vasudeva Sarma <vasudeva.sarma@sanger.ac.uk>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE
+
+*/
+
+/* The pupose of this code is to demontrate the library apis and need proper error handling and optimization */
+
+#include <getopt.h>
+#include <unistd.h>
+#include <htslib/sam.h>
+
+/// print_usage - print the demo_usage
+/** @param fp pointer to the file / terminal to which demo_usage to be dumped
+returns nothing
+*/
+static void print_usage(FILE *fp)
+{
+    fprintf(fp, "Usage: read_aux [-n position] [-i infile] [<infile>]\n\
+Dumps the aux field data from given position\n");
+}
+
+/// printauxdata - prints aux data
+/** @param fp - file to which it to be printed - stdout or null
+ *  @param type - aux type
+ *  @param idx - index in array, -1 when not an array type
+ *  @param data - data
+ *  recurses when the data is array type
+returns 1 on failure 0 on success
+*/
+int printauxdata(FILE *fp, char type, int32_t idx, const uint8_t *data)
```

Comment:
See `bam_aux_get_str` and `sam_format_aux1` APIs in htslib/sam.h.

These basically replace this with code that converts and aux tag to a printable kstring.

Although we may wish to demonstrate somewhere how to extract the value numerically and do something with it.  bam_aux2i etc.

Reply:
The intent was the later, to demonstrate the raw access of data, based on aux type - basic and array of different data types.